File: packages/shared-types/src/weather.ts
export interface WeatherPoint {
  id: string;
  latitude: number;
  longitude: number;
  temperature: number;
  humidity: number;
  windSpeed: number;
  windDirection: number;
  pressure: number;
  description: string;
  icon: string;
  timestamp: string;
  location: string;
}

export interface WeatherForecast {
  id: string;
  weatherPointId: string;
  forecastedAt: string;
  temperature: number;
  humidity: number;
  windSpeed: number;
  pressure: number;
  description: string;
  icon: string;
  timestamp: string;
}

export interface GeolocationCoords {
  latitude: number;
  longitude: number;
  accuracy: number;
}

export interface WebSocketMessage {
  type: "weather_update" | "forecast_update" | "error" | "ping" | "pong";
  data?: WeatherPoint | WeatherForecast | string;
  timestamp: string;
}

export interface WeatherApiResponse {
  current: WeatherPoint;
  forecast: WeatherForecast[];
}

export interface ForecastRequest {
  latitude: number;
  longitude: number;
  historicalData: Array<{
    timestamp: string;
    temperature: number;
  }>;
}

File: apps/api/src/models/weather.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use sqlx::FromRow;
use uuid::Uuid;

#[derive(Debug, Serialize, Deserialize, FromRow, Clone)]
pub struct WeatherPoint {
    pub id: Uuid,
    pub latitude: f64,
    pub longitude: f64,
    pub temperature: f64,
    pub humidity: f64,
    pub wind_speed: f64,
    pub wind_direction: f64,
    pub pressure: f64,
    pub description: String,
    pub icon: String,
    pub timestamp: DateTime<Utc>,
    pub location: String,
}

#[derive(Debug, Serialize, Deserialize, FromRow)]
pub struct WeatherForecast {
    pub id: Uuid,
    pub weather_point_id: Uuid,
    pub forecasted_at: DateTime<Utc>,
    pub temperature: f64,
    pub humidity: f64,
    pub wind_speed: f64,
    pub pressure: f64,
    pub description: String,
    pub icon: String,
    pub timestamp: DateTime<Utc>,
}

#[derive(Debug, Deserialize)]
pub struct WeatherRequest {
    pub latitude: f64,
    pub longitude: f64,
}

#[derive(Debug, Deserialize)]
pub struct ForecastRequest {
    pub latitude: f64,
    pub longitude: f64,
    pub historical_data: Vec<HistoricalDataPoint>,
}

#[allow(dead_code)]
#[derive(Debug, Deserialize)]
pub struct HistoricalDataPoint {
    pub timestamp: DateTime<Utc>,
    pub temperature: f64,
}
